cli/tools/clean.rs

```diff
@@ -69,6 +92,502 @@ pub fn clean(flags: Arc<Flags>) -> Result<(), AnyError> {
+
+  if !dry_run {
+    log_stats(&state, &deno_dir.root);
+
+    if let Some(dir) = node_modules_path {
+      log_stats(&node_modules_cleaned, dir);
+    }
+    if let Some(dir) = options.vendor_dir_path() {
+      log_stats(&vendor_cleaned, dir);
+    }
+  }
+
+  Ok(())
+}
+
+fn log_stats(state: &CleanState, dir: &Path) {
+  if state.bytes_removed == 0
+    && state.dirs_removed == 0
+    && state.files_removed == 0
+  {
+    return;
+  }
+  log::info!(
+    "{} {}",
+    colors::green("Removed"),
+    colors::gray(&format!(
+      "{} files, {} from {}",
+      state.files_removed + state.dirs_removed,
+      display::human_size(state.bytes_removed as f64),
+      dir.display()
+    ))
+  );
+}
+
+fn add_jsr_meta_paths(
+  graph: &ModuleGraph,
+  path_trie: &mut PathTrie,
+  jsr_url: &Url,
+  url_to_path: &dyn Fn(&Url) -> Result<PathBuf, AnyError>,
+) -> Result<(), AnyError> {
+  for package in graph.packages.mappings().values() {
+    let Ok(base_url) = jsr_url.join(&format!("{}/", &package.name)) else {
+      continue;
+    };
+    let keep = url_to_path(&base_url.join("meta.json").unwrap())?;
+    path_trie.insert(&keep);
+    let keep = url_to_path(
+      &base_url
+        .join(&format!("{}_meta.json", package.version))
+        .unwrap(),
+    )?;
+    path_trie.insert(&keep);
+  }
+  Ok(())
+}
+
+fn walk_removing(
+  state: &mut CleanState,
+  walker: WalkDir,
+  trie: &PathTrie,
+  base: &Path,
+  dry_run: bool,
+) -> Result<(), AnyError> {
+  let mut walker = walker.into_iter();
+  while let Some(entry) = walker.next() {
+    let entry = entry?;
+    if let Some(found) = trie.find(entry.path()) {
+      if entry.file_type().is_dir() && matches!(found, Found::Match) {
+        walker.skip_current_dir();
+        continue;
+      }
+      continue;
+    }
+    if !entry.path().starts_with(base) {
+      panic!("VERY BAD");
+    }
+    if entry.file_type().is_dir() {
+      if dry_run {
+        #[allow(clippy::print_stderr)]
+        {
+          eprintln!("would remove dir: {}", entry.path().display());
+        }
+      } else {
+        rm_rf(state, entry.path())?;
+      }
+      walker.skip_current_dir();
+    } else if dry_run {
+      #[allow(clippy::print_stderr)]
+      {
+        eprintln!("would remove file: {}", entry.path().display());
+      }
+    } else {
+      remove_file(state, entry.path(), Some(entry.metadata()?))?;
+    }
+  }
+
+  Ok(())
+}
+
+fn clean_node_modules(
+  state: &mut CleanState,
+  keep_pkgs: &HashSet<deno_npm::NpmPackageCacheFolderId>,
+  dir: &Path,
+  dry_run: bool,
+) -> Result<(), AnyError> {
+  if !dir.ends_with("node_modules") || !dir.is_dir() {
+    bail!("not a node_modules directory");
+  }
+  let base = dir.join(".deno");
+  if !base.exists() {
+    return Ok(());
+  }
+
+  let keep_names = keep_pkgs
+    .iter()
+    .map(deno_resolver::npm::get_package_folder_id_folder_name)
+    .collect::<HashSet<_>>();
+
+  // TODO(nathanwhit): this probably shouldn't reach directly into this code
```

Comment:
I'm not sure I understand what's the problem here - is this touching some internal implementation details?

Reply:
sort of. it's just everywhere else in the code we expose more generic methods on e.g. `NpmInstaller`. This just uses some of the code from the local installer directly. It's not incorrect, but it feels sort of wrong to me.

Reply:
We could expose it as a public API down the road?